DIPs/1NNN-RA.md

```diff
@@ -0,0 +1,409 @@
+```D
+class C : ProtoObject, Equals
+{
+    int a;
+    this(int) @safe {}
+    mixin ImplementEquals;
+}
+```
+In druntime/object.d:
+```D
+class ProtoObject {  }
+
+interface Equals
+{
+    const @nogc nothrow pure @safe scope
+    int equals(scope const ProtoObject rhs);
+}
+
+mixin template ImplementEquals(M...)
+{
+    // Equals mixin template implementation
+}
+...
+```
+
+The final points are of crucial importance:
+*  must be backwards compatible. The introduction of ProtoObject must not break code
+* The user must be allowed to chose what methods he desires to implement
+* Root objects must work in attributed code without issues. Since we, sadly, can't predict the future and know if and what attributes and qualifiers will be available in the language, this is yet another argument to have a ProtoObject with no methods.
+
+## Prior Work
+As some of you already know, both Java and C# have an `Object` class as the superclass of all classes. For both languages, `Object`
+defines a set of default methods, with Java baking into `Object` much more than C# does.
+
+#### Java
+
+Java's `java.lang.Object` defines the following methods:
+
+| Method name                                                 | Description |
+| ----------------------------------------------------------- | ------------------------------------------------------------------------------------------- |
+| boolean equals(Object o);                                   | Gives generic way to compare objects|
+| Class getClass();                                           | The Class class gives us more information about the object|
+| int hashCode();                                             | Returns a hash value that is used to search objects in a collection|
+| void notify();                                              | Used in synchronizing threads|
+| void notifyAll();                                           | Used in synchronizing threads|
+| String toString();                                          | Can be used to convert the object to String|
+| void wait();                                                | Used in synchronizing threads|
+| protected Object clone() throws CloneNotSupportedException; | Return a new object that are exactly the same as the current object|
+| protected void finalize() throws Throwable;                 | This method is called just before an object is garbage collected |
+
+Each object also has an `Object monitor` that is used in Java `synchronized` sections. Basically it is a semaphore, indicating if a critical section code is being executed by a thread or not. Before a critical section can be executed, the thread must obtain an Object monitor. Only one thread at a time can own that object's monitor.
```

Comment:
Here, on the contrary, "Object monitor" should probably not be in code font. "Monitor" is used as prose because there's no such member. Right>